File: poetry-tools/js/src/converters.ts
import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import * as zlib from 'zlib';

const readFile = promisify(fs.readFile);

/**
 * Class to convert between IPv6 addresses and poetic phrases with checksum
 */
export class IPv6PoetryConverter {
  private wordlistDir: string;
  private words: string[] = [];
  private wordToIndex: Map<string, number> = new Map();
  private initialized = false;
  
  /**
   * Create a converter
   * @param wordlistDir Directory containing the wordlist file
   */
  constructor(wordlistDir = '../../wordlists') {
    this.wordlistDir = wordlistDir;
  }
  
  /**
   * Initialize the converter by loading the wordlist
   */
  async initialize(): Promise<void> {
    if (this.initialized) return;
    
    // Load the wordlist
    const filepath = path.join(this.wordlistDir, 'wordlist.txt');
    
    try {
      const content = await readFile(filepath, 'utf8');
      this.words = content.trim().split('\n').map(word => word.trim().toLowerCase());
      
      // Verify wordlist has correct number of entries
      if (this.words.length !== 65536) {
        console.warn(`Warning: Wordlist has ${this.words.length} entries, expected 65536`);
      }
      
      // Build reverse lookup
      for (let idx = 0; idx < this.words.length; idx++) {
        this.wordToIndex.set(this.words[idx], idx);
      }
      
      this.initialized = true;
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        throw new Error(`Wordlist file not found: ${filepath}`);
      }
      throw error;
    }
  }
  
  /**
   * Normalize an IPv6 address according to RFC 5952
   * @param address IPv6 address
   * @returns Normalized IPv6 address
   */
  normalizeIPv6(address: string): string {
    // Basic validation
    if (!address.includes(':')) {
      throw new Error(`Invalid IPv6 address: ${address}`);
    }
    
    try {
      // Convert IPv6 address to binary representation (128 bits)
      let segments = address.split(':');
      let expandedSegments: string[] = [];
      
      // Handle :: notation (compressed zeroes)
      const doubleColonIndex = segments.indexOf('');
      if (doubleColonIndex !== -1) {
        // Find how many segments are missing
        const providedSegments = segments.filter(s => s !== '').length;
        const missingSegments = 8 - providedSegments;
        
        // Create an array with the correct number of zeroes
        const zeroes = Array(missingSegments).fill('0000');
        
        // Replace :: with the zeroes
        segments = [
          ...segments.slice(0, doubleColonIndex).filter(s => s !== ''),
          ...zeroes,
          ...segments.slice(doubleColonIndex + 1).filter(s => s !== '')
        ];
      }
      
      // Ensure we have 8 segments
      if (segments.length !== 8) {
        throw new Error(`Invalid IPv6 address: wrong number of segments in ${address}`);
      }
      
      // Expand each segment to 4 digits
      expandedSegments = segments.map(segment => {
        // Convert to number, then back to hex string with padded zeroes
        const num = parseInt(segment, 16);
        return num.toString(16).padStart(4, '0');
      });
      
      return expandedSegments.join(':');
    } catch (error) {
      throw new Error(`Invalid IPv6 address: ${address} - ${error}`);
    }
  }
  
  /**
   * Calculate a checksum for the address segments
   * @param decimalValues Array of decimal values for each segment
   * @returns Index for the checksum word
   */
  calculateChecksum(decimalValues: number[]): number {
    // Convert segments to bytes
    const buffer = Buffer.alloc(16); // 8 segments * 2 bytes each
    
    for (let i = 0; i < 8; i++) {
      buffer.writeUInt16BE(decimalValues[i], i * 2);
    }
    
    // Use zlib CRC32 for a simple checksum
    const checksum = zlib.crc32(buffer) & 0xFFFF; // Keep only 16 bits
    return checksum;
  }
  
  /**
   * Convert an IPv6 address to a poetic phrase with checksum
   * @param ipv6Address IPv6 address to convert
   * @returns Poetic phrase with checksum
   */
  async addressToPoetry(ipv6Address: string): Promise<string> {
    if (!this.initialized) {
      await this.initialize();
    }
    
    // Normalize the address
    const normalized = this.normalizeIPv6(ipv6Address);
    
    // Split into segments
    const segments = normalized.split(':');
    
    // Convert each segment to decimal
    const decimalValues = segments.map(segment => parseInt(segment, 16));
    
    // Map each segment to a word
    const words = decimalValues.map(value => {
      const wordIdx = value % this.words.length;
      return this.words[wordIdx];
    });
    
    // Calculate and add checksum word
    const checksumIdx = this.calculateChecksum(decimalValues);
    const checksumWord = this.words[checksumIdx];
    words.push(checksumWord);
    
    return words.join(' ');
  }
  
  /**
   * Convert a poetic phrase back to an IPv6 address
   * @param poeticPhrase Poetic phrase to convert
   * @returns IPv6 address
   */
  async poetryToAddress(poeticPhrase: string): Promise<string> {
    if (!this.initialized) {
      await this.initialize();
    }
    
    // Split the phrase into words
    const words = poeticPhrase.toLowerCase().trim().split(/\s+/);
    
    // We expect 9 words: 8 for the address + 1 checksum
    if (words.length < 8) {
      throw new Error(`Not enough words for IPv6 address. Need at least 8 words, got ${words.length}`);
    }
    
    // Extract address words (excluding checksum)
    const addressWords = words.slice(0, 8);
    
    // Convert each word to a hexadecimal segment
    const segments: string[] = [];
    const decimalValues: number[] = [];
    
    for (let i = 0; i < addressWords.length; i++) {
      const word = addressWords[i];
      const idx = this.wordToIndex.get(word);
      
      if (idx === undefined) {
        console.warn(`Warning: Word '${word}' not found in wordlist`);
        segments.push('0000');
        decimalValues.push(0);
      } else {
        const hexValue = idx.toString(16).padStart(4, '0');
        segments.push(hexValue);
        decimalValues.push(idx);
      }
    }
    
    // Verify checksum if provided
    if (words.length >= 9) {
      const checksumWord = words[8];
      const expectedChecksum = this.calculateChecksum(decimalValues);
      const expectedWord = this.words[expectedChecksum];
      
      if (checksumWord !== expectedWord) {
        console.warn(`Warning: Checksum mismatch! Expected '${expectedWord}', got '${checksumWord}'`);
        console.warn("The phrase may contain transcription errors");
      }
    }
    
    // Construct the IPv6 address
    const ipv6Address = segments.join(':');
    
    // Normalize it according to RFC 5952
    return this.normalizeIPv6(ipv6Address);
  }
}

File: poetry-tools/python/ipv6poetry/converters.py
import os
import re
import ipaddress
import zlib

class IPv6PoetryConverter:
    """Convert between IPv6 addresses and memorable poetic phrases with checksum"""
    
    def __init__(self, wordlist_dir):
        """
        Initialize converter with a single wordlist
        
        Parameters:
        wordlist_dir (str): Directory containing the wordlist file
        """
        self.wordlist_dir = wordlist_dir
        
        # Load the wordlist
        self.words = []
        self.word_to_index = {}
        
        filepath = os.path.join(wordlist_dir, "wordlist.txt")
        
        if not os.path.exists(filepath):
            raise FileNotFoundError(f"Wordlist file not found: {filepath}")
        
        with open(filepath, 'r') as f:
            self.words = [line.strip().lower() for line in f]
            
            # Build reverse lookup
            for idx, word in enumerate(self.words):
                self.word_to_index[word] = idx
        
        if len(self.words) != 65536:
            print(f"Warning: Wordlist contains {len(self.words)} words, expected 65536")
    
    def normalize_ipv6(self, address):
        """Normalize an IPv6 address according to RFC 5952"""
        try:
            # ipaddress module handles normalization
            ip = ipaddress.IPv6Address(address)
            return str(ip)
        except ValueError as e:
            raise ValueError(f"Invalid IPv6 address: {address}") from e
    
    def calculate_checksum(self, address_segments):
        """
        Calculate a checksum word for the address segments
        
        Parameters:
        address_segments (list): List of decimal values for each segment
        
        Returns:
        int: Index for the checksum word
        """
        # Convert segments to bytes
        bytes_data = b''
        for segment in address_segments:
            bytes_data += segment.to_bytes(2, byteorder='big')
        
        # Use zlib CRC32 for a simple checksum
        checksum = zlib.crc32(bytes_data) & 0xFFFF  # Keep only 16 bits
        return checksum
    
    def address_to_poetry(self, ipv6_address):
        """
        Convert an IPv6 address to a memorable poetic phrase
        
        Parameters:
        ipv6_address (str): IPv6 address to convert
        
        Returns:
        str: Memorable poetic phrase representing the address
        """
        # Normalize the address for processing
        normalized = self.normalize_ipv6(ipv6_address)
        
        # Split into segments and expand :: notation
        segments = []
        if '::' in normalized:
            parts = normalized.split('::')
            left_parts = parts[0].split(':') if parts[0] else []
            right_parts = parts[1].split(':') if parts[1] else []
            missing = 8 - len(left_parts) - len(right_parts)
            segments = left_parts + ['0'] * missing + right_parts
        else:
            segments = normalized.split(':')
        
        # Convert segments to decimal
        decimal_values = [int(segment, 16) for segment in segments]
        
        # Map each segment to a word
        words = []
        for value in decimal_values:
            # Get the corresponding word from our wordlist
            word_idx = value % len(self.words)
            words.append(self.words[word_idx])
        
        # Calculate and add checksum word
        checksum_idx = self.calculate_checksum(decimal_values)
        checksum_word = self.words[checksum_idx]
        words.append(checksum_word)
        
        # Create the final phrase
        return ' '.join(words)
    
    def poetry_to_address(self, poetic_phrase):
        """
        Convert a memorable poetic phrase back to an IPv6 address
        
        Parameters:
        poetic_phrase (str): Poetic phrase to convert
        
        Returns:
        str: IPv6 address (canonicalized)
        """
        # Normalize the input phrase
        phrase = poetic_phrase.lower().strip()
        words = phrase.split()
        
        # We expect 9 words: 8 for the address + 1 checksum
        if len(words) < 8:
            raise ValueError(f"Not enough words for IPv6 address. Need at least 8 words, got {len(words)}")
        
        # Extract address words (excluding checksum)
        address_words = words[:8]
        
        # Convert each word to a hexadecimal segment
        segments = []
        decimal_values = []
        
        for word in address_words:
            if word in self.word_to_index:
                idx = self.word_to_index[word]
                hex_value = format(idx, '04x')
                segments.append(hex_value)
                decimal_values.append(idx)
            else:
                # Word not found, try to handle gracefully
                print(f"Warning: Word '{word}' not found in wordlist")
                segments.append('0000')
                decimal_values.append(0)
        
        # Verify checksum if provided
        if len(words) >= 9:
            checksum_word = words[8]
            expected_checksum = self.calculate_checksum(decimal_values)
            expected_word = self.words[expected_checksum]
            
            if checksum_word != expected_word:
                print(f"Warning: Checksum mismatch! Expected '{expected_word}', got '{checksum_word}'")
                print("The phrase may contain transcription errors")
        
        # Construct the IPv6 address
        ipv6_address = ':'.join(segments)
        
        # Ensure address is canonical
        return self.normalize_ipv6(ipv6_address)